Provisioning - 2nd try :)

yusefnapora opened the pull request:
Basically #124 with fewer breaking changes.  Based on master instead of the streaming dev branch.

Instead of changing JournalServer.run and removing JournalServer.main, overloads JournalServer.run with one that accepts a Config and translates that into the args expected by the existing run method.

Also, puts DynamoDatastore.Config back where it was, but removes the aws credentials, so we can use whatever has been configured using amazons tools.

Vagrant / ansible changes:

Moves `Vagrantfile` and the ansible `meta` directory into the new `provisioning` directory, instead of putting them in the project root.  To run vagrant, you should either set the `VAGRANT_CWD` environment variable to `$PWD/provisioning` or use `./scripts/vagrant.sh`, which will set the variable for you and pass any args on to the real `vagrant` command.   So, e.g. `./scripts/vagrant.sh up` will start the vm.

Note that, because the Vagrantfile is not at the project root, inside the vm `/vagrant` is mapped to the `provisioning` subdirectory.  The project root is available at `/mediachain` inside the vm.

Things I'd like to do:
- [ ] pull property file parsing into the Config parsing phase, so we can optionally use a property file instead of command line args for the "unified" entry point
- [ ] try to get private networking to work with my vagrant setup and stop forwarding ports to the host machine.
- [ ] possibly add an alternate multi-machine Vagrantfile to use to spin up a test cluster.  Would obviously need more resources, so shouldn't be the default :)

yusefnapora committed:
vagrant + ansible config

yusefnapora committed:
move Vagrantfile and ansible config to provisioning subdir

- requires launching vagrant with ./scripts/vagrant.sh wrapper script
  or setting `VAGRANT_CWD` environment var to $PWD/provisioning

yusefnapora committed:
unify configuration and transactor / rpc main entrypoints

yusefnapora committed:
don't use $PWD in vagrant script

- allows calling the script from directories other than the project
  root

vyzo commented:
just needs a little whitespace trimming; :+1:

yusefnapora committed:
whitespace

vyzo committed:
add simple provisioning playbook for transactor instances

parkan commented:
@bigs could use your eye on this as resident "have used ansible before" expert

bigs commented on provisioning/meta/dev.yml:
```diff
@@ -0,0 +1,6 @@
+---
```
i guess this should be in `playbooks`

imo we should choose, `provisioning` or `meta` as a tld. i use `meta` by convention, but don't really care too much. both seems... unnecessary :stuck_out_tongue:

bigs commented on provisioning/meta/playbooks/transactor-setup.yml:
```diff
@@ -0,0 +1,20 @@
+    apt: upgrade=yes
+  - name: setup git ppa repo
+    apt_repository: repo=ppa:git-core/ppa
+  - name: install git
```
why are we not using the `git` install present in the ubuntu repo

vyzo replied:
It's very old -- 1.9.x something.
I also were getting some 404s for the deb in repo in earlier instance setups, out of which this playbook evolved.

bigs commented on provisioning/meta/playbooks/transactor-setup.yml:
```diff
@@ -0,0 +1,20 @@
+  - name: checkout mediachain
+    git: repo=https://github.com/mediachain/mediachain.git dest=/home/ubuntu/mediachain
+  - name: run setup script
+    shell: /home/ubuntu/mediachain/scripts/ubuntu-setup.sh && touch /home/ubuntu/.ubuntu-setup
```
let's adapt this to use ansible instead of being a shell script. ansible is idempotent and offers a lot of flexibility

vyzo replied:
Let's do that the next time around? I am kind of content with state of affairs as it is right now.

bigs commented on provisioning/meta/playbooks/transactor-setup.yml:
```diff
@@ -0,0 +1,20 @@
+      creates: /home/ubuntu/.ubuntu-setup
+  - name: set ownership of mediachain src tree
+    file: path=/home/ubuntu/mediachain owner=ubuntu group=ubuntu recurse=yes
+  - name: prepare /mnt/transactor
```
are we using ebs?

vyzo replied:
not for transactors, the data goes to ephemeral storage.

bigs commented on provisioning/meta/roles/dynamodb_local/meta/main.yml:
```diff
@@ -0,0 +1,16 @@
+galaxy_info:
+  author: Cole Brown <cole.brown@gmail.com>
+  description: Installs the development version of the Amazon DynamoDB server
```
we should drop a comment explaining we hijacked the one existing on ansible galaxy

yusefnapora replied:
ah yeah, good point.

bigs commented on scripts/vagrant.sh:
```diff
@@ -0,0 +1,11 @@
+
+
+SCRIPT_DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )"
+PROJECT_ROOT="$SCRIPT_DIR/.."
```
``` bash
PROJECT_ROOT=`git rev-parse --show-toplevel`
```

yusefnapora replied:
woah, cool

bigs commented:
generally fine w/ changes. few small things here and there. only major request is `ubuntu-setup.sh` becomes an ansible role

vyzo committed:
re-arrange provisioning hierarchy out of meta, move dev.yml to playbooks

yusefnapora committed:
move ansible roles dir to playbooks/roles

yusefnapora committed:
acknowledge original author of dynamodb_local role

yusefnapora committed:
define a 'transactor' role that sets up java and sbt

- the transactor role just pulls in the dependencies for running
  a transactor. so it installs java 8 and sbt, but doesn't checkout
  the source code, etc.

- dev.yml and transactor-provisioning.yml each depend on the transactor
  role with their own customizations:

  - dev.yml adds a dependency on dynamodb-local
  - transactor-provisioning.yml checks out the current mediachain code
    from git and prepares the /mnt/transactor dir for storage on ec2

yusefnapora committed:
change author of role to mediachain team
